compiler/rustc_trait_selection/src/solve/normalizes_to/mod.rs

```diff
@@ -235,14 +234,38 @@ impl<'tcx> assembly::GoalKind<'tcx> for NormalizesTo<'tcx> {
             //
             // And then map these args to the args of the defining impl of `Assoc`, going
             // from `[u32, u64]` to `[u32, i32, u64]`.
-            let impl_args_with_gat =
-                goal.predicate.alias.args.rebase_onto(tcx, goal_trait_ref.def_id, impl_args);
-            let args = ecx.translate_args(
-                goal.param_env,
-                impl_def_id,
-                impl_args_with_gat,
-                assoc_def.defining_node,
-            );
+            let args = match assoc_def.defining_node {
+                Node::Trait(_) => goal.predicate.alias.args,
+                Node::Impl(target_impl_def_id) => {
+                    let impl_args_with_gat = goal.predicate.alias.args.rebase_onto(
+                        tcx,
+                        goal_trait_ref.def_id,
+                        impl_args,
+                    );
+                    if target_impl_def_id == impl_def_id {
+                        // Same impl, no need to rebase.
+                        impl_args_with_gat
+                    } else {
+                        let target_args = ecx.fresh_args_for_item(target_impl_def_id);
+                        let target_trait_ref = tcx
+                            .impl_trait_ref(target_impl_def_id)
+                            .unwrap()
+                            .instantiate(tcx, target_args);
+                        // Relate source impl to target impl by equating trait refs.
+                        ecx.eq(goal.param_env, impl_trait_ref, target_trait_ref)?;
+                        // Also add predicates since they may be needed to constrain the
+                        // target impl's params.
+                        ecx.add_goals(
+                            GoalSource::Misc,
+                            tcx.predicates_of(target_impl_def_id)
+                                .instantiate(tcx, target_args)
+                                .into_iter()
+                                .map(|(pred, _)| goal.with(tcx, pred)),
+                        );
+                        impl_args_with_gat.rebase_onto(tcx, impl_def_id, target_args)
+                    }
+                }
+            };
```

Comment:
can you move this into a seperate method? :thinking: it's fairly involved and feels like it's also conceptionally self-contained.